src/ORM/RelatedData/StandardRelatedDataService.php

```diff
@@ -0,0 +1,507 @@
+            }
+            $results = $this->fetchManyManyResults($record, $class, $component, false);
+            $this->addResultsToClassIDs($classIDs, $results, $componentClass);
+        }
+    }
+
+    /**
+     * Query the database to retrieve many-many results
+     *
+     * @param DataObject $record - The DataObject whose usage data is being retrieved, usually a File
+     * @param string $class - example: My\App\SomePageType
+     * @param string $component - example: 'SomeFiles' - My\App\SomePageType::SomeFiles()
+     * @param bool $reverse - true: SomePage::SomeFiles(), false: SomeFile::SomePages()
+     * @return Query|null
+     */
+    private function fetchManyManyResults(
+        DataObject $record,
+        string $class,
+        string $component,
+        bool $reverse
+    ): ?Query {
+        // Example php file: class MyPage ... private static $many_many = [ 'MyFile' => File::class ]
+        $data = $this->dataObjectSchema->manyManyComponent($class, $component);
+        if (!$data || !($data['join'] ?? false)) {
+            return null;
+        }
+        $joinTableName = $this->deriveJoinTableName($data);
+        if (!ClassInfo::hasTable($joinTableName)) {
+            return null;
+        }
+        $usesThroughTable = $data['join'] != $joinTableName;
+
+        $parentField = preg_replace('#ID$#', '', $data['parentField']) . 'ID';
+        $childField = preg_replace('#ID$#', '', $data['childField']) . 'ID';
+        $selectField = !$reverse ? $childField : $parentField;
+        $selectFields = [$selectField];
+        $whereField = !$reverse ? $parentField : $childField;
+
+        // Support for polymorphic through objects such FileLink that allow for multiple class types on one side e.g.
+        // ParentID: int, ParentClass: enum('File::class, SiteTree::class, ElementContent::class, ...')
+        if ($usesThroughTable) {
+            $dbFields = $this->dataObjectSchema->databaseFields($data['join']);
+            if ($parentField === 'ParentID' && isset($dbFields['ParentClass'])) {
+                $selectFields[] = 'ParentClass';
+                if (!$reverse) {
+                    return null;
+                }
+            }
+        }
+
+        // Prevent duplicate queries which can happen when an Image is inserted on a Page subclass via TinyMCE
+        // and FileLink will make the same query multiple times for all the different page subclasses because
+        // the FileLink is associated with the Base Page class database table
+        $queryIden = implode('-', array_merge($selectFields, [$joinTableName, $whereField, $record->ID]));
+        if (array_key_exists($queryIden, $this->queryIdens)) {
+            return null;
+        }
+        $this->queryIdens[$queryIden] = true;
+
+        return SQLSelect::create(
+            sprintf('"' . implode('", "', $selectFields) . '"'),
+            sprintf('"%s"', $joinTableName),
+            sprintf('"%s" = %u', $whereField, $record->ID)
```

Comment:
You should be parameterising $record->ID to improve performance; DB layer can re-use prepared queries without baking in a hard-coded ID value.